File: .tmp_sc4013/java_22/4004_minimum_moves_to_balance_circular_array_ii.java
// CONFIG class=Solution method=minMoves types=None
// LeetCode 4004 - Minimum Moves to Balance Circular Array II
// https://leetcode.com/problems/minimum-moves-to-balance-circular-array-ii/

import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Deque;
import java.util.List;

class Solution {
    private static final int INF = 1000000000;

    static class Edge {
        int to, cap, cost, rev;
        Edge(int to, int cap, int cost, int rev) {
            this.to = to; this.cap = cap; this.cost = cost; this.rev = rev;
        }
    }

    static class MinCostMaxFlow {
        int n;
        List<Edge>[] graph;

        @SuppressWarnings("unchecked")
        MinCostMaxFlow(int n_) {
            n = n_;
            graph = new ArrayList[n];
            for (int i = 0; i < n; i++) graph[i] = new ArrayList<>();
        }

        void addEdge(int u, int v, int cap, int cost) {
            graph[u].add(new Edge(v, cap, cost, graph[v].size()));
            graph[v].add(new Edge(u, 0, -cost, graph[u].size() - 1));
        }

        long minCostFlow(int source, int sink, int maxFlow) {
            long totalCost = 0;
            int currentFlow = 0;
            while (currentFlow < maxFlow) {
                int[] dist = new int[n], parentNode = new int[n], parentEdge = new int[n];
                boolean[] inQueue = new boolean[n];
                Arrays.fill(dist, INF);
                Arrays.fill(parentNode, -1);
                Arrays.fill(parentEdge, -1);
                Deque<Integer> q = new ArrayDeque<>();
                q.add(source);
                dist[source] = 0;
                inQueue[source] = true;
                while (!q.isEmpty()) {
                    int u = q.poll();
                    inQueue[u] = false;
                    for (int i = 0; i < graph[u].size(); i++) {
                        Edge e = graph[u].get(i);
                        if (e.cap > 0 && dist[e.to] > dist[u] + e.cost) {
                            dist[e.to] = dist[u] + e.cost;
                            parentNode[e.to] = u;
                            parentEdge[e.to] = i;
                            if (!inQueue[e.to]) {
                                inQueue[e.to] = true;
                                q.add(e.to);
                            }
                        }
                    }
                }
                if (dist[sink] == INF) return -1;
                int pushFlow = maxFlow - currentFlow;
                for (int cur = sink; cur != source; cur = parentNode[cur]) {
                    Edge e = graph[parentNode[cur]].get(parentEdge[cur]);
                    if (e.cap < pushFlow) pushFlow = e.cap;
                }
                for (int cur = sink; cur != source; cur = parentNode[cur]) {
                    int p = parentNode[cur];
                    int idx = parentEdge[cur];
                    int rev = graph[p].get(idx).rev;
                    graph[p].get(idx).cap -= pushFlow;
                    graph[cur].get(rev).cap += pushFlow;
                }
                currentFlow += pushFlow;
                totalCost += (long) pushFlow * dist[sink];
            }
            return totalCost;
        }
    }

    public long minMoves(int[] balance) {
        int totalBalance = 0, totalDeficit = 0;
        for (int x : balance) {
            totalBalance += x;
            if (x < 0) totalDeficit += -x;
        }
        if (totalBalance < 0) return -1;
        if (totalDeficit == 0) return 0;
        int n = balance.length;
        int source = n, sink = n + 1;
        MinCostMaxFlow mcmf = new MinCostMaxFlow(n + 2);
        for (int i = 0; i < n; i++) {
            int x = balance[i];
            if (x > 0) mcmf.addEdge(source, i, x, 0);
            else if (x < 0) mcmf.addEdge(i, sink, -x, 0);
            mcmf.addEdge(i, (i + 1) % n, INF, 1);
            mcmf.addEdge(i, (i - 1 + n) % n, INF, 1);
        }
        return mcmf.minCostFlow(source, sink, totalDeficit);
    }
}

File: .tmp_rs4013/rust_check22/4004_minimum_moves_to_balance_circular_array_ii.rs
struct Solution;
// LeetCode 4004 - Minimum Moves to Balance Circular Array II
// https://leetcode.com/problems/minimum-moves-to-balance-circular-array-ii/

use std::collections::VecDeque;

const INF: i32 = 1_000_000_000;

struct Edge {
    to: usize,
    cap: i32,
    cost: i32,
    rev: usize,
}

struct MinCostMaxFlow {
    n: usize,
    graph: Vec<Vec<Edge>>,
}

impl MinCostMaxFlow {
    fn new(n: usize) -> Self {
        Self {
            n,
            graph: (0..n).map(|_| Vec::new()).collect(),
        }
    }

    fn add_edge(&mut self, u: usize, v: usize, cap: i32, cost: i32) {
        let rev_v = self.graph[v].len();
        let rev_u = self.graph[u].len();
        self.graph[u].push(Edge {
            to: v,
            cap,
            cost,
            rev: rev_v,
        });
        self.graph[v].push(Edge {
            to: u,
            cap: 0,
            cost: -cost,
            rev: rev_u,
        });
    }

    fn min_cost_flow(&mut self, source: usize, sink: usize, max_flow: i32) -> i64 {
        let mut total_cost = 0i64;
        let mut current_flow = 0;
        while current_flow < max_flow {
            let mut dist = vec![INF; self.n];
            let mut parent_node = vec![-1i32; self.n];
            let mut parent_edge = vec![-1i32; self.n];
            let mut in_queue = vec![false; self.n];
            let mut q = VecDeque::new();
            q.push_back(source);
            dist[source] = 0;
            in_queue[source] = true;
            while let Some(u) = q.pop_front() {
                in_queue[u] = false;
                for i in 0..self.graph[u].len() {
                    let e_to = self.graph[u][i].to;
                    let e_cap = self.graph[u][i].cap;
                    let e_cost = self.graph[u][i].cost;
                    if e_cap > 0 && dist[e_to] > dist[u] + e_cost {
                        dist[e_to] = dist[u] + e_cost;
                        parent_node[e_to] = u as i32;
                        parent_edge[e_to] = i as i32;
                        if !in_queue[e_to] {
                            in_queue[e_to] = true;
                            q.push_back(e_to);
                        }
                    }
                }
            }
            if dist[sink] == INF {
                return -1;
            }
            let mut push_flow = max_flow - current_flow;
            let mut cur = sink;
            while cur != source {
                let p = parent_node[cur] as usize;
                let idx = parent_edge[cur] as usize;
                if self.graph[p][idx].cap < push_flow {
                    push_flow = self.graph[p][idx].cap;
                }
                cur = p;
            }
            cur = sink;
            while cur != source {
                let p = parent_node[cur] as usize;
                let idx = parent_edge[cur] as usize;
                let rev = self.graph[p][idx].rev;
                self.graph[p][idx].cap -= push_flow;
                self.graph[cur][rev].cap += push_flow;
                cur = p;
            }
            current_flow += push_flow;
            total_cost += push_flow as i64 * dist[sink] as i64;
        }
        total_cost
    }
}

impl Solution {
    pub fn min_moves(balance: Vec<i32>) -> i64 {
        let mut total_balance = 0;
        let mut total_deficit = 0;
        for &x in &balance {
            total_balance += x;
            if x < 0 {
                total_deficit += -x;
            }
        }
        if total_balance < 0 {
            return -1;
        }
        if total_deficit == 0 {
            return 0;
        }
        let n = balance.len();
        let source = n;
        let sink = n + 1;
        let mut mcmf = MinCostMaxFlow::new(n + 2);
        for i in 0..n {
            let x = balance[i];
            if x > 0 {
                mcmf.add_edge(source, i, x, 0);
            } else if x < 0 {
                mcmf.add_edge(i, sink, -x, 0);
            }
            mcmf.add_edge(i, (i + 1) % n, INF, 1);
            mcmf.add_edge(i, (i + n - 1) % n, INF, 1);
        }
        mcmf.min_cost_flow(source, sink, total_deficit)
    }
}

fn main() {}
